db/hints/manager.cc

```diff
@@ -267,6 +281,16 @@ sync_point::shard_rps manager::calculate_current_sync_point(std::span<const gms:
 }
 
 future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_rps& rps) {
+    // During the migration process, endpoint managers are all stopped. They're resumed
+    // by the end of it, so in case we call this function while hinted handoff is being
+    // migrated to host IDs, we need to wait for it to end to avoid race conditions.
+    //
+    // This scenario is quite unlikely because when stopping an endpoint manager,
+    // the sync points should all be invalidated and exceptions should be issued,
+    // but it's better to be safe. The danger mostly lies in the part where we're iterating
+    // over endpoint managers.
+    co_await _migrating_done.get_future();
```

Comment:
You're right. But if `coroutine::parallel_for` is not preemptive, then shouldn't this work? Thanks to this await, it's guaranteed that when we proceed to the next instruction in the function, there is no ongoing migration. Once we start the lambda for all endpoint managers, we're safe. If a migration starts afterwards, but before the coroutine finishes, all hint senders will be stopped and `wait_until_hints_are_replayed_up_to()` will return an exception indicating an abort.

Reply:
Yes, I also think that it will work _right now_, but relying on such non-local assumptions makes the code fragile. Somebody might change that code later and the code here will break. Relying on explicit synchronization primitives is better because they give a stronger indication to the people that they should think twice before removing them - people don't give as much attention to (non)preemptiveness.